server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java

```diff
@@ -0,0 +1,207 @@
+        DiscoveryNode mostRecentNonNullMaster = null;
+        for (int i = masterHistory.size() - 1; i >= 0; i--) {
+            TimeAndMaster timeAndMaster = masterHistory.get(i);
+            if (timeAndMaster.master != null) {
+                mostRecentNonNullMaster = timeAndMaster.master;
+                break;
+            }
+        }
+        return mostRecentNonNullMaster;
+    }
+
+    /**
+     * Returns true if for the life of this MasterHistory (30 minutes) only one non-null node has been master, and the master has switched
+     * from that node to null n times.
+     * @param n The number of times the non-null master must have switched to null
+     * @return True if there has been a single non-null master and it has switched to null n or more times.
+     */
+    public boolean hasSameMasterGoneNullNTimes(int n) {
+        if (getDistinctMastersSeen().size() != 1) {
+            return false;
+        }
+        boolean seenNonNull = false;
+        int timesMasterHasGoneNull = 0;
+        for (TimeAndMaster timeAndMaster : masterHistory) {
+            if (timeAndMaster.master != null) {
+                seenNonNull = true;
+            } else if (seenNonNull) {
+                timesMasterHasGoneNull++;
+            }
+        }
+        return timesMasterHasGoneNull >= n;
+    }
+
+    /**
+     * Returns the set of distinct non-null master nodes seen in this history.
+     * @return The set of all non-null master nodes seen. Could be empty
+     */
+    public Set<DiscoveryNode> getDistinctMastersSeen() {
+        removeOldMasterHistory();
+        return masterHistory.stream()
+            .filter(timeAndMaster -> timeAndMaster.master != null)
+            .map(timeAndMaster -> timeAndMaster.master)
+            .collect(Collectors.toSet());
+    }
+
+    /**
+     * Returns true if a non-null master was seen at any point in the last n seconds, or if the last-seen master was more than n seconds
+     * ago and non-null.
+     * @param n The number of seconds to look back
+     * @return true if the current master is non-null or if a non-null master was seen in the last n seconds
+     */
+    public boolean hasSeenMasterInLastNSeconds(int n) {
+        long now = nowSupplier.get();
+        long nSecondsAgo = now - (n * 1000L);
+        return getCurrentMaster() != null
+            || masterHistory.stream().anyMatch(timeAndMaster -> timeAndMaster.time > nSecondsAgo && timeAndMaster.master != null);
+    }
+
+    /**
+     * Clears out anything from masterHistory that is from more than 30 minutes before now (but leaves the newest entry in even if it is
+     * more than 30 minutes old). Rather than being scheduled, this method is called from other methods in this class as needed
+     */
+    private void removeOldMasterHistory() {
+        if (masterHistory.size() < 2) {
+            return;
+        }
+        long now = nowSupplier.get();
+        long thirtyMinutesAgo = now - (30 * 60 * 1000);
```

Comment:
`TimeValue` can be used to express various units ( https://github.com/elastic/elasticsearch/blob/master/libs/core/src/main/java/org/elasticsearch/core/TimeValue.java#L410 )

Reply:
Done.